File: nextjs-api/scripts/bulk_download.mjs
#!/usr/bin/env node
/**
 * bulk_download.mjs — Bulk vessel image downloader (Node.js)
 *
 * Usage:
 *   node scripts/bulk_download.mjs                                  # defaults
 *   node scripts/bulk_download.mjs --csv ships.csv --out images/    # custom
 *   node scripts/bulk_download.mjs --concurrency 3 --delay 2        # tuning
 */

import { createReadStream, mkdirSync, writeFileSync } from "node:fs";
import { createInterface } from "node:readline";
import { parseArgs } from "node:util";
import { join } from "node:path";

const { values: args } = parseArgs({
  options: {
    csv: { type: "string", default: "ships.csv" },
    out: { type: "string", default: "images" },
    url: { type: "string", default: "http://127.0.0.1:3000" },
    concurrency: { type: "string", default: "2" },
    retries: { type: "string", default: "3" },
    delay: { type: "string", default: "1" },
    report: { type: "string", default: "report.csv" },
  },
});

const BASE_URL = args.url.replace(/\/$/, "");
const CONCURRENCY = parseInt(args.concurrency);
const RETRIES = parseInt(args.retries);
const DELAY = parseFloat(args.delay) * 1000;
const OUT_DIR = args.out;
const REPORT_PATH = args.report;

// ── Read CSV ────────────────────────────────────────────────────────────────

async function readCsv(path) {
  const vessels = [];
  const rl = createInterface({ input: createReadStream(path, "utf-8") });
  let headers = null;

  for await (const raw of rl) {
    const line = raw.replace(/^\uFEFF/, "").trim();
    if (!line) continue;
    const cols = line.split(",").map((c) => c.trim());
    if (!headers) {
      headers = cols.map((h) => h.toLowerCase());
      continue;
    }
    const row = Object.fromEntries(headers.map((h, i) => [h, cols[i] ?? ""]));
    const mmsi = row.mmsi?.trim();
    const name = (row.name || row.vessel_name || "").trim();
    if (mmsi && name) vessels.push({ mmsi, name });
  }
  return vessels;
}

// ── Filename helpers ────────────────────────────────────────────────────────

function safeName(name) {
  return name.replace(/[^\w-]/g, "_").replace(/_+/g, "_").replace(/^_|_$/g, "");
}

const CT_TO_EXT = {
  "image/jpeg": ".jpg",
  "image/png": ".png",
  "image/webp": ".webp",
  "image/avif": ".avif",
  "image/gif": ".gif",
};

// ── Download one vessel ─────────────────────────────────────────────────────

async function downloadOne(vessel) {
  const url = `${BASE_URL}/api/vessel/image?mmsi=${vessel.mmsi}&name=${encodeURIComponent(vessel.name)}`;
  let lastErr = "";

  for (let attempt = 1; attempt <= RETRIES; attempt++) {
    const t0 = performance.now();
    try {
      const res = await fetch(url, { signal: AbortSignal.timeout(30_000) });
      const elapsed = performance.now() - t0;

      if (res.ok) {
        const ct = (res.headers.get("content-type") ?? "image/jpeg").split(";")[0].trim();
        const ext = CT_TO_EXT[ct] || ".jpg";
        const filename = `${safeName(vessel.name)}_${vessel.mmsi}${ext}`;
        const filepath = join(OUT_DIR, filename);
        const buf = Buffer.from(await res.arrayBuffer());
        writeFileSync(filepath, buf);
        return {
          success: true,
          filepath,
          sizeBytes: buf.length,
          elapsedMs: elapsed,
          cached: res.headers.get("x-cache") === "HIT",
          error: null,
        };
      }
      lastErr = `HTTP ${res.status}`;
    } catch (err) {
      lastErr = err.message?.slice(0, 120) ?? "unknown";
    }

    if (attempt < RETRIES) {
      await new Promise((r) => setTimeout(r, 1000 * attempt));
    }
  }

  return { success: false, filepath: null, sizeBytes: 0, elapsedMs: 0, cached: false, error: lastErr };
}

// ── Main ────────────────────────────────────────────────────────────────────

async function main() {
  const vessels = await readCsv(args.csv);
  if (vessels.length === 0) {
    console.log(`  No vessels found in ${args.csv}. Expected CSV with columns: mmsi, name`);
    process.exit(1);
  }

  mkdirSync(OUT_DIR, { recursive: true });
  console.log(`  Loaded ${vessels.length} vessels from ${args.csv}`);

  // Health check
  process.stdout.write(`\n  Checking API at ${BASE_URL} ... `);
  try {
    const r = await fetch(`${BASE_URL}/api/health`, { signal: AbortSignal.timeout(5000) });
    if (r.ok) console.log("OK");
    else console.log(`WARNING: HTTP ${r.status}`);
  } catch (err) {
    console.log(`\n  Cannot reach API: ${err.message}`);
    console.log("  Start the server first: npm run dev");
    process.exit(1);
  }

  const etaMin = (vessels.length * DELAY / 1000) / 60;
  console.log(`  Downloading ${vessels.length} vessel images (concurrency=${CONCURRENCY}, delay=${DELAY/1000}s)`);
  console.log(`  Estimated time: ~${etaMin.toFixed(0)} minutes\n`);

  const results = [];
  const total = vessels.length;
  let doneCount = 0;
  const startTime = performance.now();
  const pad = String(total).length;

  // Semaphore via simple queue
  let running = 0;
  const queue = [...vessels];
  const allDone = [];

  async function worker() {
    while (queue.length > 0) {
      const vessel = queue.shift();
      if (!vessel) break;
      const result = await downloadOne(vessel);
      doneCount++;
      const icon = result.success ? "OK" : "FAIL";
      const elapsed = (performance.now() - startTime) / 1000;
      const rate = doneCount / elapsed;
      console.log(
        `  [${String(doneCount).padStart(pad)}/${total}] ${icon.padEnd(4)}  ${vessel.name.padEnd(30)}  MMSI=${vessel.mmsi}  ${result.elapsedMs.toFixed(0).padStart(7)}ms  (${rate.toFixed(1)}/s)`
      );
      results.push({ vessel, ...result });
      await new Promise((r) => setTimeout(r, DELAY));
    }
  }

  const workers = Array.from({ length: CONCURRENCY }, () => worker());
  await Promise.all(workers);

  // Report CSV
  const reportLines = [
    "mmsi,name,status,filepath,size_bytes,elapsed_ms,error",
    ...results.map((r) =>
      [
        r.vessel.mmsi,
        r.vessel.name,
        r.success ? "OK" : "FAIL",
        r.filepath ?? "",
        r.sizeBytes,
        r.elapsedMs.toFixed(0),
        r.error ?? "",
      ].join(",")
    ),
  ];
  writeFileSync(REPORT_PATH, reportLines.join("\n") + "\n");
  console.log(`\n  Report written to ${REPORT_PATH}`);

  // Summary
  const ok = results.filter((r) => r.success).length;
  const fail = total - ok;
  const totalBytes = results.reduce((s, r) => s + r.sizeBytes, 0);
  const wallTime = (performance.now() - startTime) / 1000;

  console.log(`\n  ${"=".repeat(56)}`);
  console.log(`  BULK DOWNLOAD COMPLETE`);
  console.log(`  ${"=".repeat(56)}`);
  console.log(`  Total vessels:   ${total}`);
  console.log(`  Successful:      ${ok}  (${((ok / total) * 100).toFixed(1)}%)`);
  console.log(`  Failed:          ${fail}`);
  console.log(`  Total size:      ${(totalBytes / 1024 / 1024).toFixed(1)} MB`);
  console.log(`  Wall time:       ${wallTime.toFixed(1)}s`);
  console.log(`  Throughput:      ${(total / wallTime).toFixed(1)} vessels/s`);
  console.log(`  ${"=".repeat(56)}\n`);

  if (fail > 0) {
    console.log(`  Failed vessels:`);
    for (const r of results.filter((r) => !r.success)) {
      console.log(`    ${r.vessel.name.padEnd(30)}  MMSI=${r.vessel.mmsi}  error=${r.error}`);
    }
  }
}

main();

File: bulk_download.py
"""
bulk_download.py — Bulk vessel image downloader
-------------------------------------------------
Reads a CSV of vessels (mmsi, name) and downloads images via the
Vessel Image API, saving them with meaningful filenames.

Usage:
    python bulk_download.py                                # defaults
    python bulk_download.py --csv ships.csv --out images/  # custom
    python bulk_download.py --concurrency 10 --retries 3   # tuning
    python bulk_download.py --url http://remote:8000       # remote API

CSV format (header required):
    mmsi,name
    311052100,ANASTASIA K
    235113366,QUEEN MARY 2

Output:
    images/ANASTASIA_K_311052100.jpg
    images/QUEEN_MARY_2_235113366.jpg
    report.csv  (success/fail log for every vessel)
"""

from __future__ import annotations

import argparse
import asyncio
import csv
import os
import re
import sys
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

import httpx

# Fix Windows console encoding
if sys.platform == "win32":
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
    sys.stderr.reconfigure(encoding="utf-8", errors="replace")


# ── Models ───────────────────────────────────────────────────────────────────

@dataclass
class Vessel:
    mmsi: str
    name: str

    @property
    def safe_name(self) -> str:
        """Filesystem-safe version of the vessel name."""
        return re.sub(r"[^\w\-]", "_", self.name.strip()).strip("_")

    @property
    def filename_stem(self) -> str:
        return f"{self.safe_name}_{self.mmsi}"


@dataclass
class DownloadResult:
    vessel: Vessel
    success: bool
    filepath: Optional[str] = None
    size_bytes: int = 0
    elapsed_ms: float = 0
    error: Optional[str] = None
    cached: bool = False


# ── CSV reader ───────────────────────────────────────────────────────────────

def read_vessels_csv(path: str) -> list[Vessel]:
    vessels = []
    with open(path, newline="", encoding="utf-8-sig") as f:
        reader = csv.DictReader(f)
        # Normalize header names
        if reader.fieldnames:
            reader.fieldnames = [h.strip().lower() for h in reader.fieldnames]
        for row in reader:
            mmsi = (row.get("mmsi") or "").strip()
            name = (row.get("name") or row.get("vessel_name") or "").strip()
            if mmsi and name:
                vessels.append(Vessel(mmsi=mmsi, name=name))
    return vessels


# ── Extension from content-type ──────────────────────────────────────────────

_CT_TO_EXT = {
    "image/jpeg": ".jpg",
    "image/png": ".png",
    "image/webp": ".webp",
    "image/avif": ".avif",
    "image/gif": ".gif",
}


def _ext_from_ct(content_type: str) -> str:
    ct = content_type.split(";")[0].strip().lower()
    return _CT_TO_EXT.get(ct, ".jpg")


# ── Single download ─────────────────────────────────────────────────────────

async def download_one(
    client: httpx.AsyncClient,
    base_url: str,
    vessel: Vessel,
    out_dir: Path,
    retries: int,
) -> DownloadResult:
    url = f"{base_url}/vessel/image"
    params = {"mmsi": vessel.mmsi, "name": vessel.name}

    last_err = ""
    for attempt in range(1, retries + 1):
        t0 = time.perf_counter()
        try:
            r = await client.get(url, params=params, timeout=30.0)
            elapsed = (time.perf_counter() - t0) * 1000

            if r.status_code == 200:
                ct = r.headers.get("content-type", "image/jpeg")
                ext = _ext_from_ct(ct)
                filename = f"{vessel.filename_stem}{ext}"
                filepath = out_dir / filename
                filepath.write_bytes(r.content)
                cached = r.headers.get("X-Cache", "") == "HIT"
                return DownloadResult(
                    vessel=vessel,
                    success=True,
                    filepath=str(filepath),
                    size_bytes=len(r.content),
                    elapsed_ms=elapsed,
                    cached=cached,
                )
            else:
                last_err = f"HTTP {r.status_code}"
        except httpx.TimeoutException:
            elapsed = (time.perf_counter() - t0) * 1000
            last_err = "TIMEOUT"
        except Exception as exc:
            elapsed = (time.perf_counter() - t0) * 1000
            last_err = str(exc)[:120]

        # Back off before retry
        if attempt < retries:
            await asyncio.sleep(1.0 * attempt)

    return DownloadResult(
        vessel=vessel,
        success=False,
        elapsed_ms=elapsed,
        error=last_err,
    )


# ── Bulk orchestrator ────────────────────────────────────────────────────────

async def bulk_download(
    vessels: list[Vessel],
    base_url: str,
    out_dir: Path,
    concurrency: int,
    retries: int,
) -> list[DownloadResult]:
    out_dir.mkdir(parents=True, exist_ok=True)
    results: list[DownloadResult] = []
    semaphore = asyncio.Semaphore(concurrency)
    total = len(vessels)
    done_count = 0
    start_time = time.perf_counter()

    async def _worker(vessel: Vessel) -> DownloadResult:
        nonlocal done_count
        async with semaphore:
            result = await download_one(client, base_url, vessel, out_dir, retries)
        done_count += 1
        icon = "OK" if result.success else "FAIL"
        elapsed_total = time.perf_counter() - start_time
        rate = done_count / elapsed_total if elapsed_total > 0 else 0
        print(
            f"  [{done_count:>{len(str(total))}}/{total}] "
            f"{icon:4s}  {vessel.name:30s}  MMSI={vessel.mmsi}  "
            f"{result.elapsed_ms:>7.0f}ms  "
            f"({rate:.1f}/s)",
        )
        return result

    async with httpx.AsyncClient(
        follow_redirects=True,
        limits=httpx.Limits(max_connections=concurrency + 10, max_keepalive_connections=concurrency),
    ) as client:
        # Health check
        print(f"\n  Checking API at {base_url} ...", end=" ", flush=True)
        try:
            r = await client.get(f"{base_url}/health", timeout=5.0)
            if r.status_code == 200:
                print(f"OK")
            else:
                print(f"WARNING: HTTP {r.status_code}")
        except Exception as exc:
            print(f"\n  Cannot reach API: {exc}")
            print("  Start the server first: uvicorn main:app --host 0.0.0.0 --port 8000")
            sys.exit(1)

        print(f"  Downloading {total} vessel images (concurrency={concurrency}, retries={retries})\n")

        tasks = [asyncio.create_task(_worker(v)) for v in vessels]
        results = await asyncio.gather(*tasks)

    return list(results)


# ── Report ───────────────────────────────────────────────────────────────────

def write_report(results: list[DownloadResult], report_path: Path):
    with open(report_path, "w", newline="", encoding="utf-8") as f:
        writer = csv.writer(f)
        writer.writerow(["mmsi", "name", "status", "filepath", "size_bytes", "elapsed_ms", "error"])
        for r in results:
            writer.writerow([
                r.vessel.mmsi,
                r.vessel.name,
                "OK" if r.success else "FAIL",
                r.filepath or "",
                r.size_bytes,
                f"{r.elapsed_ms:.0f}",
                r.error or "",
            ])


def print_summary(results: list[DownloadResult], elapsed_s: float):
    total = len(results)
    ok = sum(1 for r in results if r.success)
    fail = total - ok
    total_bytes = sum(r.size_bytes for r in results)

    print(f"\n  {'='*56}")
    print(f"  BULK DOWNLOAD COMPLETE")
    print(f"  {'='*56}")
    print(f"  Total vessels:   {total}")
    print(f"  Successful:      {ok}  ({ok/total*100:.1f}%)" if total else "")
    print(f"  Failed:          {fail}")
    print(f"  Total size:      {total_bytes / 1024 / 1024:.1f} MB")
    print(f"  Wall time:       {elapsed_s:.1f}s")
    print(f"  Throughput:      {total / elapsed_s:.1f} vessels/s" if elapsed_s > 0 else "")
    print(f"  {'='*56}")

    if fail > 0:
        print(f"\n  Failed vessels:")
        for r in results:
            if not r.success:
                print(f"    {r.vessel.name:30s}  MMSI={r.vessel.mmsi}  error={r.error}")
    print()


# ── CLI ──────────────────────────────────────────────────────────────────────

def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Bulk download vessel images via the API")
    p.add_argument("--csv", default="ships.csv", help="Path to CSV file with mmsi,name columns (default: ships.csv)")
    p.add_argument("--out", default="images", help="Output directory for images (default: images/)")
    p.add_argument("--url", default="http://127.0.0.1:8000", help="Base URL of Vessel Image API")
    p.add_argument("--concurrency", type=int, default=5, help="Max concurrent downloads (default: 5)")
    p.add_argument("--retries", type=int, default=3, help="Retries per vessel on failure (default: 3)")
    p.add_argument("--report", default="report.csv", help="Output CSV report path (default: report.csv)")
    return p.parse_args()


if __name__ == "__main__":
    args = parse_args()

    vessels = read_vessels_csv(args.csv)
    if not vessels:
        print(f"  No vessels found in {args.csv}. Expected CSV with columns: mmsi, name")
        sys.exit(1)

    print(f"  Loaded {len(vessels)} vessels from {args.csv}")

    t0 = time.perf_counter()
    results = asyncio.run(
        bulk_download(
            vessels=vessels,
            base_url=args.url.rstrip("/"),
            out_dir=Path(args.out),
            concurrency=args.concurrency,
            retries=args.retries,
        )
    )
    elapsed = time.perf_counter() - t0

    report_path = Path(args.report)
    write_report(results, report_path)
    print(f"  Report written to {report_path}")

    print_summary(results, elapsed)
